product_variant_exclusion/models/product.py

```diff
@@ -0,0 +1,174 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from openerp import models, fields, api, _
+from openerp.exceptions import ValidationError, UserError
+import psycopg2
+
+
+class ProductTemplate(models.Model):
+
+    _inherit = 'product.template'
+
+    variant_exclusion_ids = fields.One2many(
+        'product.variant.exclusion', 'product_tmpl_id', 'Variants exclusion'
+    )
+
+    allowed_value_ids = fields.Many2many('product.attribute.value',
+                                         compute='_compute_allowed_values')
+
+    @api.depends('attribute_line_ids.value_ids')
+    def _compute_allowed_values(self):
+        for template in self:
+            template.allowed_value_ids = template.attribute_line_ids.mapped(
+                'value_ids')
+
+    @api.multi
+    def write(self, vals):
+        res = super(ProductTemplate, self).write(vals)
+        if 'variant_exclusion_ids' in vals:
+            self.create_variant_ids()
+            self.delete_variant_ids()
+        return res
+
+    @api.multi
+    def delete_variant_ids(self):
+        exclusions = self.env['product.variant.exclusion'].search(
+            [('product_tmpl_id', '=', self.id)])
+        for template in self:
+            for product in template.product_variant_ids:
+                for excl in exclusions:
+                    if set(excl.attribute_value_ids.ids) <= set(
+                            product.attribute_value_ids.ids):
+                        try:
+                            with self.env.cr.savepoint():
+                                product.unlink()
+                        except (psycopg2.Error, ValidationError):
+                            product.write({'active': False})
+                        break
+
+
+class ProductAttributeLine(models.Model):
+
+    _inherit = 'product.attribute.line'
+
+    @api.multi
+    def write(self, vals):
+        attr_values = vals.get('value_ids', False) and vals.get(
+            'value_ids')[0][2]
+        if attr_values and len(attr_values) <= 1:
+            for line in self:
+                for exclusion in line.product_tmpl_id.variant_exclusion_ids:
+                    if set(attr_values) <= set(
+                            exclusion.attribute_value_ids.ids):
+                        raise UserError(_(
+                            'You cannot leave only one attribute value on a '
```

Comment:
This needs a usability test but I think its OK.  The scenario I'm thinking of is removing some attributes and adding others. 

But also the logic looks a bit wrong/brittle.  The obvious question is why not, if an attribute is only excluded in limited circumstances (i.e. dependent on another attribute value), why do we need this error? Would it not make more sense to unlink the corresponding attribute values of the exclusions (e.g. if we deleted Male from gender and size 40 was excluded from Female, unlink 40)?  But also I can see a scenario where due to a combination of exclusions this error wouldn't be triggered when it should. (when len(attr_values) > 1 but the remaining attr values are also part of an exclusion themself)